python311Packages.python-lsp-black: build failure, failing tests

### Describe the bug

As of master @ caacb878cdc69d7f6addfe8b526d0b664261b576, this package is failing to build.

This seems related to https://github.com/NixOS/nixpkgs/commit/f8e8b40db295aae1215cc8a38f6cae7502bdfb47

A clear and concise description of what the bug is.

### Steps To Reproduce
Steps to reproduce the behavior:
1. `nix-build -A python311Packages.python-lsp-black`

Result:

```

Executing pytestCheckPhase
============================= test session starts ==============================
platform linux -- Python 3.11.8, pytest-8.0.2, pluggy-1.4.0
rootdir: /build/source
collected 21 items                                                             

tests/test_plugin.py ....F......F.........                               [100%]

=================================== FAILURES ===================================
___________________ test_pylsp_format_document_syntax_error ____________________

src = '    x = 1+2\n', dst = 'x = 1 + 2\n'

    def assert_equivalent(src: str, dst: str) -> None:
        """Raise AssertionError if `src` and `dst` aren't equivalent."""
        try:
>           src_ast = parse_ast(src)

/nix/store/fdbzd7fz5p3qdvpw8a4212lcp13p9nzv-python3.11-black-24.3.0/lib/python3.11/site-packages/black/__init__.py:1526: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

src = '    x = 1+2\n'

    def parse_ast(src: str) -> ast.AST:
        # TODO: support Python 4+ ;)
        versions = [(3, minor) for minor in range(3, sys.version_info[1] + 1)]
    
        first_error = ""
        for version in sorted(versions, reverse=True):
            try:
                return _parse_single_version(src, version, type_comments=True)
            except SyntaxError as e:
                if not first_error:
                    first_error = str(e)
    
        # Try to parse without type comments
        for version in sorted(versions, reverse=True):
            try:
                return _parse_single_version(src, version, type_comments=False)
            except SyntaxError:
                pass
    
>       raise SyntaxError(first_error)
E       SyntaxError: unexpected indent (<unknown>, line 1)

/nix/store/fdbzd7fz5p3qdvpw8a4212lcp13p9nzv-python3.11-black-24.3.0/lib/python3.11/site-packages/black/parsing.py:148: SyntaxError

The above exception was the direct cause of the following exception:

config = <pylsp.config.config.Config object at 0x7ffff5632990>
invalid_document = <pylsp.workspace.Document object at 0x7ffff563fc10>

    def test_pylsp_format_document_syntax_error(config, invalid_document):
>       result = pylsp_format_document(config, invalid_document)

tests/test_plugin.py:179: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pylsp_black/plugin.py:38: in pylsp_format_document
    return format_document(config, document)
pylsp_black/plugin.py:73: in format_document
    formatted_text = format_text(text=text, config=config, lines=lines)
pylsp_black/plugin.py:114: in format_text
    formatted_text = black.format_file_contents(
/nix/store/fdbzd7fz5p3qdvpw8a4212lcp13p9nzv-python3.11-black-24.3.0/lib/python3.11/site-packages/black/__init__.py:1083: in format_file_contents
    check_stability_and_equivalence(
/nix/store/fdbzd7fz5p3qdvpw8a4212lcp13p9nzv-python3.11-black-24.3.0/lib/python3.11/site-packages/black/__init__.py:1057: in check_stability_and_equivalence
    assert_equivalent(src_contents, dst_contents)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

src = '    x = 1+2\n', dst = 'x = 1 + 2\n'

    def assert_equivalent(src: str, dst: str) -> None:
        """Raise AssertionError if `src` and `dst` aren't equivalent."""
        try:
            src_ast = parse_ast(src)
        except Exception as exc:
>           raise ASTSafetyError(
                "cannot use --safe with this file; failed to parse source file AST: "
                f"{exc}\n"
                "This could be caused by running Black with an older Python version "
                "that does not support new syntax used in your source file."
            ) from exc
E           black.parsing.ASTSafetyError: cannot use --safe with this file; failed to parse source file AST: unexpected indent (<unknown>, line 1)
E           This could be caused by running Black with an older Python version that does not support new syntax used in your source file.

/nix/store/fdbzd7fz5p3qdvpw8a4212lcp13p9nzv-python3.11-black-24.3.0/lib/python3.11/site-packages/black/__init__.py:1528: ASTSafetyError
_____________________ test_pylsp_format_range_syntax_error _____________________

src = '    x = 1+2\n', dst = 'x = 1 + 2\n'

    def assert_equivalent(src: str, dst: str) -> None:
        """Raise AssertionError if `src` and `dst` aren't equivalent."""
        try:
>           src_ast = parse_ast(src)

/nix/store/fdbzd7fz5p3qdvpw8a4212lcp13p9nzv-python3.11-black-24.3.0/lib/python3.11/site-packages/black/__init__.py:1526: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

src = '    x = 1+2\n'

    def parse_ast(src: str) -> ast.AST:
        # TODO: support Python 4+ ;)
        versions = [(3, minor) for minor in range(3, sys.version_info[1] + 1)]
    
        first_error = ""
        for version in sorted(versions, reverse=True):
            try:
                return _parse_single_version(src, version, type_comments=True)
            except SyntaxError as e:
                if not first_error:
                    first_error = str(e)
    
        # Try to parse without type comments
        for version in sorted(versions, reverse=True):
            try:
                return _parse_single_version(src, version, type_comments=False)
            except SyntaxError:
                pass
    
>       raise SyntaxError(first_error)
E       SyntaxError: unexpected indent (<unknown>, line 1)

/nix/store/fdbzd7fz5p3qdvpw8a4212lcp13p9nzv-python3.11-black-24.3.0/lib/python3.11/site-packages/black/parsing.py:148: SyntaxError

The above exception was the direct cause of the following exception:

config = <pylsp.config.config.Config object at 0x7ffff5634950>
invalid_document = <pylsp.workspace.Document object at 0x7ffff567f610>

    def test_pylsp_format_range_syntax_error(config, invalid_document):
        range = {"start": {"line": 0, "character": 0}, "end": {"line": 1, "character": 0}}
    
>       result = pylsp_format_range(config, invalid_document, range=range)

tests/test_plugin.py:253: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pylsp_black/plugin.py:46: in pylsp_format_range
    return format_document(config, document, range)
pylsp_black/plugin.py:73: in format_document
    formatted_text = format_text(text=text, config=config, lines=lines)
pylsp_black/plugin.py:114: in format_text
    formatted_text = black.format_file_contents(
/nix/store/fdbzd7fz5p3qdvpw8a4212lcp13p9nzv-python3.11-black-24.3.0/lib/python3.11/site-packages/black/__init__.py:1083: in format_file_contents
    check_stability_and_equivalence(
/nix/store/fdbzd7fz5p3qdvpw8a4212lcp13p9nzv-python3.11-black-24.3.0/lib/python3.11/site-packages/black/__init__.py:1057: in check_stability_and_equivalence
    assert_equivalent(src_contents, dst_contents)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

src = '    x = 1+2\n', dst = 'x = 1 + 2\n'

    def assert_equivalent(src: str, dst: str) -> None:
        """Raise AssertionError if `src` and `dst` aren't equivalent."""
        try:
            src_ast = parse_ast(src)
        except Exception as exc:
>           raise ASTSafetyError(
                "cannot use --safe with this file; failed to parse source file AST: "
                f"{exc}\n"
                "This could be caused by running Black with an older Python version "
                "that does not support new syntax used in your source file."
            ) from exc
E           black.parsing.ASTSafetyError: cannot use --safe with this file; failed to parse source file AST: unexpected indent (<unknown>, line 1)
E           This could be caused by running Black with an older Python version that does not support new syntax used in your source file.

/nix/store/fdbzd7fz5p3qdvpw8a4212lcp13p9nzv-python3.11-black-24.3.0/lib/python3.11/site-packages/black/__init__.py:1528: ASTSafetyError
=============================== warnings summary ===============================
tests/test_plugin.py:8
  /build/source/tests/test_plugin.py:8: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
    import pkg_resources

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========================== short test summary info ============================
FAILED tests/test_plugin.py::test_pylsp_format_document_syntax_error - black.parsing.ASTSafetyError: cannot use --safe with this file; failed to p...
FAILED tests/test_plugin.py::test_pylsp_format_range_syntax_error - black.parsing.ASTSafetyError: cannot use --safe with this file; failed to p...
=================== 2 failed, 19 passed, 1 warning in 0.34s ====================
/nix/store/v5lsd029lz5lfhamivbgqyp3zdv94ah2-stdenv-linux/setup: line 1578: pop_var_context: head of shell_variables not a function context
error: builder for '/nix/store/sf03vzbryvdd6bc1rpc638b2znij2xir-python3.11-python-lsp-black-2.0.0.drv' failed with exit code 1;
       last 10 log lines:
       > tests/test_plugin.py:8
       >   /build/source/tests/test_plugin.py:8: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
       >     import pkg_resources
       >
       > -- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
       > =========================== short test summary info ============================
       > FAILED tests/test_plugin.py::test_pylsp_format_document_syntax_error - black.parsing.ASTSafetyError: cannot use --safe with this file; failed to p...
       > FAILED tests/test_plugin.py::test_pylsp_format_range_syntax_error - black.parsing.ASTSafetyError: cannot use --safe with this file; failed to p...
       > =================== 2 failed, 19 passed, 1 warning in 0.34s ====================
       > /nix/store/v5lsd029lz5lfhamivbgqyp3zdv94ah2-stdenv-linux/setup: line 1578: pop_var_context: head of shell_variables not a function context
       For full logs, run 'nix log /nix/store/sf03vzbryvdd6bc1rpc638b2znij2xir-python3.11-python-lsp-black-2.0.0.drv'.
```
### Expected behavior

Should build

### Notify maintainers

<!--
Please @ people who are in the `meta.maintainers` list of the offending package or module.
If in doubt, check `git blame` for whoever last touched something.
-->

@bryango

@cpcloud 

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.7.12, NixOS, 24.05 (Uakari), 24.05.20240403.fd281bd`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.2`
 - nixpkgs: `/nix/store/n2g5cqwv8qf5p6vjxny6pg3blbdij12k-source`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/development/python-modules/python-lsp-black/default.nix
+++ b/pkgs/development/python-modules/python-lsp-black/default.nix
@@ -24,13 +24,19 @@ buildPythonPackage rec {
   };
 
   patches =
-    /** fix test failure with black<24.2.0;
-        remove this patch once python-lsp-black>2.0.0 */
+    /** includes a series of patches fixing tests not yet released as 2.0.1+ version
+        they are meant to keep up to date with black releases */
     lib.optional
-      (with lib; (versionOlder version "2.0.1") && (versionAtLeast black.version "24.2.0"))
+      (with lib; versionAtLeast black.version "24.2.0")
       (fetchpatch {
-        url = "https://patch-diff.githubusercontent.com/raw/python-lsp/python-lsp-black/pull/56.patch";
+        url = "https://github.com/python-lsp/python-lsp-black/commit/d43b41431379f9c9bb05fab158c4d97e6d515f8f.patch";
         hash = "sha256-38bYU27+xtA8Kq3appXTkNnkG5/XgrUJ2nQ5+yuSU2U=";
+      })
+    ++ lib.optional
+      (with lib; versionAtLeast black.version "24.3.0")
+      (fetchpatch {
+        url = "https://github.com/python-lsp/python-lsp-black/commit/9298585a9d14d25920c33b188d79e820dc98d4a9.patch";
+        hash = "sha256-4u0VIS7eidVEiKRW2wc8lJVkJwhzJD/M+uuqmTtiZ7E=";
       })
     ++ [ ];
 

--- a/pkgs/development/python-modules/python-lsp-black/default.nix
+++ b/pkgs/development/python-modules/python-lsp-black/default.nix
@@ -37,8 +37,7 @@ buildPythonPackage rec {
       (fetchpatch {
         url = "https://github.com/python-lsp/python-lsp-black/commit/9298585a9d14d25920c33b188d79e820dc98d4a9.patch";
         hash = "sha256-4u0VIS7eidVEiKRW2wc8lJVkJwhzJD/M+uuqmTtiZ7E=";
-      })
-    ++ [ ];
+      });
 
   nativeBuildInputs = [
     setuptools

